django/db/models/fields/composite.py

```diff
@@ -0,0 +1,179 @@
+from django.db.models.signals import class_prepared
+from django.utils.functional import cached_property
+
+
+class TupleExact(Exact):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, Cols):
+            raise ValueError(
+                "The left-hand side of the 'exact' lookup must be an instance of Cols"
+            )
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(
+                "The right-hand side of the 'exact' lookup must be a tuple or a list"
+            )
+        if len(self.lhs) != len(self.rhs):
+            raise ValueError(
+                "The left-hand side and right-hand side of the 'exact' lookup must "
+                "have the same number of elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, WhereNode
+
+        cols = self.lhs.get_source_expressions()
+        exprs = [Exact(col, val) for col, val in zip(cols, self.rhs)]
+
+        return compiler.compile(WhereNode(exprs, connector=AND))
+
+
+class TupleIn(In):
+    def get_prep_lookup(self):
+        if not isinstance(self.lhs, Cols):
+            raise ValueError(
+                "The left-hand side of the 'in' lookup must be an instance of Cols"
+            )
+        if not isinstance(self.rhs, (tuple, list)):
+            raise ValueError(
+                "The right-hand side of the 'in' lookup must be a tuple or a list"
+            )
+        if not all(isinstance(vals, (tuple, list)) for vals in self.rhs):
+            raise ValueError(
+                "The right-hand side of the 'in' lookup must be a set of "
+                "tuples or lists"
+            )
+        if not all(len(self.lhs) == len(vals) for vals in self.rhs):
+            raise ValueError(
+                "The left-hand side and right-hand side of the 'in' lookup must "
+                "have the same number of elements"
+            )
+
+        return super().get_prep_lookup()
+
+    def as_sql(self, compiler, connection):
+        from django.db.models.sql.where import AND, OR, WhereNode
+
+        exprs = []
+        cols = self.lhs.get_source_expressions()
+
+        for vals in self.rhs:
+            exprs.append(
+                WhereNode(
+                    [Exact(col, val) for col, val in zip(cols, vals)], connector=AND
+                )
+            )
+
+        return compiler.compile(WhereNode(exprs, connector=OR))
+
+
+class Cols(Expression):
+    def __init__(self, alias, targets, output_field):
+        super().__init__(output_field=output_field)
+        self.alias, self.targets = alias, targets
+
+    def get_source_expressions(self):
+        return [Col(self.alias, target) for target in self.targets]
+
+    def set_source_expressions(self, exprs):
+        assert all(isinstance(expr, Col) for expr in exprs)
+        assert len(exprs) == len(self.targets)
```

Comment:
Shouldn't this perform assigments? Or maybe it's meant to happen during clone relabeling?

Reply:
I'm not sure, but as far as I've seen, this function is only called from the `resolve_expression` function. I'll check more in detail as I'm not sure about the significance of this function, but it looks like it works without performing any assignments yes.

Reply:
already resolved 👍